docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,864 @@
+ * optimally upload and downloads to and from S3.
+ * <p>
+ * The list of features includes:
+ * <ul>
+ *   <li>Parallel uploads and downloads</li>
+ *   <li>Bandwidth limiting</li>
+ *   <li>Pause and resume of transfers</li>
+ * </ul>
+ * <p>
+ * <b>Usage Example:</b>
+ * <pre>
+ * {@code
+ * // Create using all default configuration values
+ * S3TransferManager tm = S3TranferManager.create();
+ *
+ * // Using custom configuration values to set max upload speed to avoid
+ * // saturating the network interface.
+ * S3TransferManager tm = S3TransferManager.builder()
+ *         .configuration(TransferManagerConfiguration.builder()
+ *             .maxUploadBytesSecond(32 * 1024 * 1024) // 32 MiB
+ *             .build()
+ *         .build();
+ * }
+ * </pre>
+ */
+public interface S3TransferManager {
+    /**
+     * Download an object identified by the bucket and key from S3 to the given
+     * file.
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate transfer
+     * Download myFileDownload = tm.downloadObject(BUCKET, KEY, Paths.get("/tmp/myFile.txt");
+     * // Wait for transfer to complete
+     * myFileDownload().completionFuture().join();
+     * }
+     * </pre>
+     *
+     */
+    default Download downloadObject(String bucket, String key, Path file) {
+        return downloadObject(DownloadObjectRequest.builder()
+                        .getObjectRequest(GetObjectRequest.builder()
+                                .bucket(bucket)
+                                .key(key)
+                                .build())
+                        .build(),
+                file);
+    }
+
+    /**
+     * Download an object using an S3 presigned URL to the given file.
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate transfer
+     * Download myFileDownload = tm.downloadObject(myPresignedUrl, Paths.get("/tmp/myFile.txt");
+     * // Wait for transfer to complete
+     * myFileDownload()completionFuture().join();
+     * }
+     * </pre>
+     */
+    default Download downloadObject(URL presignedUrl, Path file) {
+        return downloadObject(DownloadObjectRequest.builder()
+                        .presignedUrl(presignedUrl)
+                        .build(),
+                file);
+    }
+    /**
+     * Download an object in S3 to the given file.
+     *
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate the transfer
+     * Download myDownload = tm.downloadObject(DownloadObjectRequest.builder()
+     *         .getObjectRequest(GetObjectRequest.builder()
+     *                 .bucket(BUCKET)
+     *                 .key(KEY)
+     *                 .build())
+     *          // Set the known length of the object to avoid a HeadObject call
+     *         .knownProperties(KnownObjectProperties.builder()
+     *                 .size(1024 * 1024 * 5)
+     *                 .build())
```

Comment:
Does this need to be a separate configuration? Can't it just be `DownloadObjectRequest.builder().size(...).build()`? We could always use an interface for it if customers want to share code.

Reply:
Yeah I think it can just be made part of the request.